bin/node-template/node/src/command.rs

```diff
@@ -184,11 +189,32 @@ pub fn run() -> sc_cli::Result<()> {
 				let task_manager =
 					sc_service::TaskManager::new(config.tokio_handle.clone(), registry)
 						.map_err(|e| sc_cli::Error::Service(sc_service::Error::Prometheus(e)))?;
+
+				let info_provider = |_, maybe_prev_info: Option<(InherentData, Digest)>| async {
```

Comment:
Slightly unfortunate that chains implementing try-runtime have to add this snippet. Also, this kinda breaks https://github.com/paritytech/substrate/issues/12975, any alternatives?

Reply:
I definitely agree that this complicates things, but I'm afraid that as long as we want to remain chain-agnostic, similar approach must be taken.

Firstly, please notice that all other subcommands of `try-runtime` were oblivious to the chain they were working with (as long it was exposing the proper runtime API calls). However, when it comes to block production (like in `fast-forward`), we are facing the process that is often unique to every chain (here: inherents and digests). I suppose that majority of the chains in the ecosystem will have exactly the same configuration here (timestamp + aura), but this won't be true everywhere.

It can be also anticipated, that sooner or later we might enrich `try-runtime` with new functionalities that will have some chain-dependent components. This is why in [12975](https://github.com/paritytech/substrate/issues/12975 I proposed splitting `try-runtime` tool into a part that contains pure (chain-agnostic) logic and a part where customization takes place. In this sense, I don't see using user snippet as breaking.

> _It seems like this is a misunderstanding. Everything in the try-runtime-cli crate is also "client side code", and has full access to system time in order to build the inherent and all. Only things within state_machine_call are actually runtime calls._

Yes, the misunderstanding comes from my poor wording - by 'client side code' I was referring to the CLI part, i.e. the code that handles user's input to the `try-runtime` tool.

Reply:
Okay, two thoughts for now:

1. it might help both of us find an alternative solution if you actually create a companion PR for this PR to Aleph-node, Polkadot, or some other client that actually requires a different setup. This will prove your point that we NEED this customization, but might shed some light into how else it could be done.
2. If not useful, then I am happy to merge this and we shall find a better solution, even if it exists, in #12975. We should just document this label properly so that other parachain teams can easily follow-along and update their client integration.

Reply:
I did not really check the code in here, but most of the stuff you are doing here is already provided for manual-seal: https://github.com/paritytech/substrate/blob/4e383428ffc6d4eba173ca257c1aaf53d1ec339c/client/consensus/manual-seal/src/consensus/aura.rs#L75-L90

In general, manual-seal already provides functionality to "fast-forward" as well.

Reply:
manual seal lives with the full client, I don't think we can reuse the code here. Nonetheless, perhaps we can reuse some of the code.

Reply:
You can for sure share logic as what I have linked is already doing what these changes are here trying to do.

Reply:
do you mean just digest computation from inherents, or a bigger part?

Reply:
while semantically `create_digest` is perfectly applicable in this PR, it requires a type that supports `AuxStore + ProvideRuntimeApi<B> + UsageProvider<B>` for Aura or `AuxStore + HeaderBackend<B> + HeaderMetadata<B, Error = sp_blockchain::Error> + UsageProvider<B> + ProvideRuntimeApi<B>` for Babe, which goes far out of scope what we use in either `try-runtime` itself or even `commands` modules

Reply:
I mean the logic should be shared. I also need to rewrite sp-api that you can use it here instead of calling functions by their name.

Reply:
Okay, apologies for misunderstanding. Do you think that I should make some steps towards this logic sharing in this PR? If so, then what would be the good place to put common things in?

Reply:
Maybe for now you just copy the relevant code to `try-build`? And make it reusable for people? So that they don't always need to implement the logic?

Reply:
I couldn't find any `try-build`, so I assumed that you meant `try-runtime`... I moved both trait and two most common (I guess) impementations to a new module within `try-runtime`. It cost a few `sp-*` dependencies, but indeed, ease of usage is way higher. Please let me know if this is what you suggested.